cpp/ql/test/query-tests/Critical/MemoryFreed/test_free.cpp

```diff
@@ -0,0 +1,229 @@
+void *malloc(int);
+void free(void *);
+bool condition();
+void use(void*);
+void *realloc(void*, unsigned long);
+int strlen(char*);
+int asprintf(char ** strp, const char * fmt, ...);
+
+
+void* test_double_free1(int *a) {
+    free(a); // GOOD
+    a[5] = 5; // BAD
+    *a = 5; // BAD
+    free(a); // BAD
+    a = (int*) malloc(8);
+    free(a); // GOOD
+    a = (int*) malloc(8);
+    if (!a) free(a);
+    return a;
+}
+
+void test_double_free_aliasing(void *a, void* b) {
+    free(a); // GOOD
+    a = b;
+    free(a); // GOOD
+    free(b); // BAD [NOT DETECTED]
+}
+
+void test_dominance1(void *a) {
+    free(a);
+    if (condition()) free(a); // BAD
+}
+
+void test_dominance2(void *a) {
+    free(a);
+    if (condition()) a = malloc(10);
+    if (condition()) free(a); // BAD
+}
+
+void test_post_dominance1(int *a)
+{
+    if (condition()) free(a); // GOOD
+    if (condition()) a[2] = 5; // GOOD
+    if (condition()) free(a); // GOOD
```

Comment:
Agreed. I think they were added as `GOOD` because the query isn't designed to find these since neither of those satisfy the post-dominator/dominator condition. I'll change them to `BAD [NOT DETECTED]`.

Reply:
Fixed in 31b71ea163736e96175ab56bbb064ae4384275e4.

Reply:
Fixed in 31b71ea163736e96175ab56bbb064ae4384275e4.